### PR TITLE
[Javascript] Added support for commonjs in nodejs package

### PR DESCRIPTION
**Pull request by Tarjei400**
No package should enforce upon creator a need to migrate to ESM. As per this pull request, I added webpack configuration that produces cjs and mjs files Correspoding to CommonJS and ESM module. After this pull request if will be up to a project which one to use.

Cheers!

### CODE REVIEW

**Comment by ericvergnaud** [comment]
Hey, nobody's enforcing anything, and we welcome contributors. Let's see how this works for other devs looking for CommonJS support.

**Comment by NilsBaumgartner1994** [comment]
@Tarjei400 thank you :-) will have a look. @ericvergnaud thanks for notification.

**Comment by Tarjei400** [comment]
@enessoylu I just mean that at the moment library requires ESM in a project and this is "enforcing" ESM in a project, I didnt mean anyone per se :P Anyways thanks for taking a look, I think it will be a usefull addition for this package!

**Comment by jharris4** [comment]
I think adding `exports` to the `package.json` is a step in the right direction.

Following down this path should help with the TypeScript target as well. For example as of TypeScript 5.0 you can do the following with `exports` in your `package.json` to export separate types for node and web:

```
{
  "exports": {
    ".": {
      "node": {
        "types": "src/index.node.d.ts",
        "default": "dist/bundle.node.js"
      },
      "browser": {
        "types": "src/index.web.d.ts",
        "default": "dist/bundle.web.js"
      }
    }
  }
}
```

And with TypeScript 5 the package entry point & types can be toggled in the `tsconfig.json` by changing `compilerOptions.moduleResolution` between `nodenext` and `bundler`.

[TypeScript docs]( https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#moduleresolution-bundler)

[NodeJS docs](https://nodejs.org/api/packages.html#conditional-exports)

**Comment by Tarjei400** [comment]
@jharris4 Sorry it took some time. I checked that export solution and added tests for nodejs esm/commonjs

File: runtime/JavaScript/spec/imports/NodeCommonJSImportSpec.cjs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+"use strict";
+var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
```
Comment (ericvergnaud) [inline runtime/JavaScript/spec/imports/NodeCommonJSImportSpec.cjs:2]: I could use an explanation of the code in this file. Why can't we just write "require('antlr4')" ?

**Comment by ericvergnaud** [comment]
If I understand correctly, this PR would generate esm and cjs for web and node, without touching any runtime code ? Sounds promising! Awaiting for my inline comments to be addressed.

File: runtime/JavaScript/spec/imports/NodeCommonJSImportSpec.cjs (hunk 1 above)
Comment (Tarjei400) [reply runtime/JavaScript/spec/imports/NodeCommonJSImportSpec.cjs:2]: Right, I just transpiled it with ts, but you are right require will do the job. Changed :+1:

File: runtime/JavaScript/webpack.config.js [hunk 2]
```diff
@@ -5,68 +5,63 @@ import {fileURLToPath} from "url";
 const __filename = fileURLToPath(import.meta.url);
 const __dirname = path.dirname(__filename);
 
-const nodeConfig = {
+
+const buildConfig = ( platform, extentions ) => ({
```
[the remark this answers is not in the document]
Comment (Tarjei400) [reply runtime/JavaScript/webpack.config.js:9]: Type, fixed

**Comment by Tarjei400** [comment]
Correct @ericvergnaud :D Addressed your comments!

**Comment by ericvergnaud** [comment]
Good stuff! Thanks so much, I know for sure this will make many developers happy.
@parrt blessed

**Comment by Tarjei400** [comment]
Myself included :D Also thanks for this great tool!

**Comment by ericvergnaud** [comment]
Fixes #4211

**Comment by jedwards1211** [comment]
@Tarjei400 big thanks for this 😊

**Comment by NilsBaumgartner1994** [comment]
NiceBeste Grüße,Nils BaumgartnerAm 10.11.2023 um 09:43 schrieb Andy Edwards ***@***.***>:﻿
@Tarjei400 big thanks for this 😊

**Pull request merged**